components/style/properties/properties.mako.rs

```diff
@@ -3085,9 +3090,338 @@ pub fn cascade(
+                    ).map(|(untyped, maybe_typed, inherits)| {
+                        has_uninherited = has_uninherited || !inherits;
+                        if let Some(typed) = maybe_typed {
+                            computed_to_insert_typed.push((name.clone(), typed));
+                        }
+                        untyped
+                    })
+                },
+                BorrowedExtraData::Precomputed(ref computed_value) => {
+                    // This specified value came from an animation or an
+                    // inherited typed custom property. We don't need to
+                    // compute!
+                    if let Some(ref registration) = registered_property_set.get(name) {
+                        has_uninherited = has_uninherited || !registration.inherits;
+                        computed_to_insert_typed.push((name.clone(), (*computed_value).clone()));
+                    }
+                    Ok(value)
+                },
+                BorrowedExtraData::InheritedUntyped => {
+                    // The specified value was inherited from an untyped custom
+                    // property. We shouldn't insert any typed value here.
+                    Ok(value)
+                },
+            }
+        },
+        /* handle_invalid */ &mut |name| {
+            registered_property_set
+                .get(name)
+                .and_then(|registration| {
+                    if !registration.inherits {
+                        registration.initial_value.as_ref()
+                            .map(|x| {
+                                let typed =
+                                    x.to_computed_value(context)
+                                    .expect("The initial value should never fail to compute.");
+                                let untyped = (&typed).into();
+                                invalid_to_insert_typed.push((name.clone(), typed));
+                                untyped
+                            })
+                    } else {
+                        None
+                    }
+                })
+                .or_else(|| inherited.as_ref().and_then(|x| x.get(name).cloned()))
+                .ok_or(())
+        }
+    );
+
+    for (name, typed) in computed_to_insert_typed {
+        computed.insert_typed(&name, None, typed);
+    }
+    for (name, typed) in invalid_to_insert_typed {
+        computed.insert_typed(&name, None, typed);
+    }
+    if has_uninherited {
+        computed.set_has_uninherited();
+    }
+}
+
+/// Determine and then compute those custom properties we want to compute before
+/// early custom properties (might avoid doing any work if we haven't declared
+/// any properties and there are no uninherited custom properties).
+/// Returns the custom properties' specified nad computed values, as well as
+/// whether or not the declaration for font-size is involved in a cycle.
+fn compute_early_custom_properties<'a, 'b: 'a, I>(
```

Comment:
If this doesn't need mako code, I'd really appreciate to have it outside of this file. I'm trying to remove it eventually.

Reply:
I think this should be in the same file that `cascade` is in as a separate function, in order to maintain modularity. If the functionality of `cascade` independent of mako code is eventually moved somewhere else, it should move with it, I think.